### ISSUE TITLE
[FEATURE] Optimize upgrade `image preloading`

### ISSUE THREAD

**Issue by w13915984028** [harvester/harvester#3059]
**Is your feature request related to a problem? Please describe.**

Currently, the `image preloading` will happen in `NODEs` one by one, for massive-node cluster, it takes long time.

**Describe the solution you'd like**

Preload image to nodes in parallel.
Or allow user to set up an image registry outside of cluster, images will be fetched on demand.
Also consider to allow user to select  `one-by-one`, `multi-in-parallel`, `fetch-on-demand-from-registry` mode.

**Describe alternatives you've considered**

**Additional context**

example of long waiting list:
![image](https://user-images.githubusercontent.com/31133476/198296206-e1cef72d-a597-4133-adfb-2ad8a1f69131.png)

**Comment by Martin-Weiss** [comment]
See also https://github.com/harvester/harvester/issues/3058

**Comment by w13915984028** [comment]
duplicated with https://github.com/harvester/harvester/issues/3058

**Comment by starbops** [comment]
I'd like to reopen this one since we're currently working on upgrade-related enhancement, and this is one of the essential bits. Plus, the scope of this issue is more specific in contrast to #3058.

**Comment by starbops** [comment]
We're adding the ability to skip the image-preload phase during an upgrade (PR #6037 is still under review) by annotating the Upgrade CR with a specific key-value pair. However, as @w13915984028 suggested, we could consolidate these two and provide a new option/setting `image-preload-behavior` to instruct the controller about the image-preload behavior during upgrades (require the controller pods to restart):

- skip (the scope covered by #6028)
- sequential (like the good old days)
- parallel (the scope of this issue)

Or is just providing another annotation, say `harvesterhci.io/parallel-image-preload: true` for the Upgrade CR good enough? cc @bk201 @brandboat

**Comment by brandboat** [comment]
> Or is just providing another annotation, say harvesterhci.io/parallel-image-preload: true for the Upgrade CR good enough? cc @bk201 @brandboat

Sorry for the late response, IMO the image-preload strategy should be a field in spec, add more annotations here is a bit mess... what if `harvesterhci.io/parallel-image-preload: true`, `harvesterhci.io/skip-image-preload: true` both exist, which one take precedence ? I propose we introduce a new field `image-preload-strategy` to upgrade CRD, doing this could avoid the problem mentioned as above.

**Comment by starbops** [comment]
@brandboat Thank you for the feedback. I was thinking of adding a new Harvester setting called `image-preload-behavior` which has three available options:

- `skip`, meaning "do not preload any image"
- `sequential`, meaning "the image-preloading will happen one node at a time." Just like the good old days.
- `parallel`, meaning "all nodes will start to preload images at the same time"

Behind the scenes, it just controls the `concurrency` variable of the `prepare` plan object.

About the user experience, they only need to select the desired behavior before initiating the upgrade: no pod restart, no upgrade object patching. Does that make sense?

**Comment by Martin-Weiss** [comment]
Just as a side note - maybe the usage of spegel might also be an option? If we would use that one - we might not need to pre-load the images to all nodes from the ISO..?

**Comment by brandboat** [comment]
> Just as a side note - maybe the usage of spegel might also be an option? If we would use that one - we might not need to pre-load the images to all nodes from the ISO..?

Thanks @Martin-Weiss, I believe https://github.com/harvester/harvester/issues/4842 address your point.

**Comment by brandboat** [comment]
> @brandboat Thank you for the feedback. I was thinking of adding a new Harvester setting called image-preload-behavior

@starbops , sounds good 😃

**Comment by w13915984028** [comment]
Like https://github.com/harvester/harvester/pull/5984/files#diff-e66e0227d36702ad7777ebc499cce7f2e8d60efda0bebbd73c14b326971b87a4R31
... `harvesterhci.io/maintain-mode-strategy`

please consider replace `image-preload-behavior` with `image-preload-strategy`, thanks.

**Comment by ibrokethecloud** [comment]
rke2 now embeds spegel registry https://github.com/spegel-org/spegel/

we should may be investigate this as this will eliminate the need for image loading as images would be pulled from the first node only?

**Comment by irishgordo** [comment]
@starbops I had a quick validation question :smile: 

For:
[step 3 in test plan](https://github.com/harvester/harvester/pull/6181)

I'm noticing:
```
 Warning  Failed                  12s (x3 over 51s)  kubelet            Failed to pull image "rancher/harvester-upgrade:master-0bde441-head": rpc error: code = NotFound desc = failed to pull and unpack image "docker.io/rancher/harvester-upgrade:master-0bde441-head": failed to resolve reference "docker.io/rancher/harvester-upgrade:master-0bde441-head": docker.io/rancher/harvest
er-upgrade:master-0bde441-head: not found
```
Trying to upgrade to master-head same iso

Should I implement the retag, as on one node I notice?:
```go
node5-test:/home/rancher # crictl image ls | grep -ie "harvester-upgrade"
docker.io/rancher/harvester-upgrade                                         0bde4419                                    a9371280e13c3       431MB
docker.io/rancher/harvester-upgrade                                         master-0bde4419-head                        a9371280e13c3       431MB
docker.io/rancher/harvester-upgrade                                         master-head                                 a9371280e13c3       431MB
```

Something like?:
```
docker tag rancher/harvester-upgrade:master-0bde4419-head rancher/harvester-upgrade:master-0bde441-head
```

But would I run into?:
- https://github.com/harvester/harvester/issues/2108

So should I not use "master-head" for testing this and instead use?:
- https://github.com/harvester/harvester/releases/tag/v1.4.0-dev-20240918
Or the new release of today:
- https://github.com/harvester/harvester/releases/tag/v1.4.0-dev-20240924 (?) when it's released?

**Comment by starbops** [comment]
@irishgordo I think this deserves its own ticket. I'll create a new one and link to the comment here.

Edit: We can wait until RC1 is out to prevent the image retagging and cluster repo pod roll-out. If you prefer to continue the upgrade, here's some notes for you:

---

Retagging should do the trick. And, yes, you're probably going to run into #2108. You can do the following to extract the tarball which contains the harvester-cluster-repo container image from the ISO image:

```
isoinfo -J -i harvester-master-amd64.iso -x /bundle/harvester/images/harvester-repo-images-master.tar.zst > harvester-repo-images-master.tar.zst
```

Upload the tarball to all nodes and import it with:

```
# on each harvester node
zstdcat harvester-repo-images-master.tar.zst | ctr -n k8s.io images import --no-unpack -
```

Finally, restart the harvester-cluster-repo deployment or just kill the pod to let the newly spawned pod use the new image.

**Comment by irishgordo** [comment]
@starbops I'm encountering a bit of a validation issue.
When trying to upgrade to the same version of Harvester, it seems that the `harvester-network-controller-manager` is stuck at deployment ready 0/1 on a 2 node cluster:

```
╭─mike at suse-workstation-team-harvester in ~
╰─○ kubectl get bundle -A
NAMESPACE     NAME                                          BUNDLEDEPLOYMENTS-READY   STATUS
fleet-local   fleet-agent-local                             1/1                       
fleet-local   local-managed-system-agent                    1/1                       
fleet-local   mcc-harvester                                 0/1                       NotReady(1) [Cluster fleet-local/local]; deployment.apps harvester-system/harvester [progressing] Available: 1/2; deployment.apps harvester-system/harvester-network-controller-manager [progressing] Deployment does not have minimum availability., Available: 1/2; deployment.apps harvester-system/harvester-webhook [progressing] Available: 1/2
fleet-local   mcc-harvester-crd                             1/1                       
fleet-local   mcc-local-managed-system-upgrade-controller   1/1                       
fleet-local   mcc-rancher-logging-crd                       1/1                       
fleet-local   mcc-rancher-monitoring-crd                    1/1   
```

Any ideas?
![Screenshot from 2024-10-07 12-31-23](https://github.com/user-attachments/assets/1e33192a-d828-444e-bf35-248ec8971c25)

[supportbundle_b5f37a65-549a-4700-9544-fe0863fec276_2024-10-07T19-31-49Z.zip](https://github.com/user-attachments/files/17283650/supportbundle_b5f37a65-549a-4700-9544-fe0863fec276_2024-10-07T19-31-49Z.zip)

**Comment by irishgordo** [comment]
Actually got past the issue with the bundle :smile:

**Comment by irishgordo** [comment]
@starbops I had a small validation question.
Is it to be expected that:
- user modifies upgrade-config
- user upgrades to version N of Harvester
- nodes as they reboot, come back online for the Harvester cluster, once RKE2 is available, then looking at the `settings/upgrade-config` it's reset to default?

Is it to be expected that the upgrade-config setting will not persist after upgrade?  That it will, instead, always reset back to "sequential" / default?

Additionally, is the UI portion of this going to be released with v1.4.0 too, or will users only be able to edit the setting from kubectl?

If that's the case I'm good closing this out :smile: :+1: 

### Environment
* Harvester
  * Version: v1.4.0-rc1
  * Profile: Bare-Metal, HP DL160 Gen9s
  * ui-source: Auto


### Verification
### 2 Node Cluster
1. :green_circle: `upgrade-config` w/ Sequential
   
   <details>
   <summary> <i>Given</i> upgrade-config setting is set to sequential/default, the upgrade will complete as expected and there will only be ever 1 prepare pod running
   </summary>

    ![Screenshot from 2024-10-07 23-31-29](https://github.com/user-attachments/assets/e43215d5-709e-4120-b2ed-8180d086833e)
    ![Screenshot from 2024-10-07 23-15-46](https://github.com/user-attachments/assets/263eeb99-2fc6-4abd-80dc-fcf50d98da21)
    ![Screenshot from 2024-10-07 23-15-27](https://github.com/user-attachments/assets/542ced2d-da66-4134-a7d4-6792c77f2965) 


     </details>
2. :green_circle: `upgrade-config` w/ Parallel
   
   <details>
   <summary> <i>Given</i> upgrade-config setting is set to parallel with a given concurrency the upgrade should succeed and the prepare pods should be equal to the N number of concurrency specified
   </summary>

    
    ![Screenshot from 2024-10-08 14-02-34](https://github.com/user-attachments/assets/7a8d0669-599d-485f-85b5-23e28663364b)
    ![Screenshot from 2024-10-08 13-51-58](https://github.com/user-attachments/assets/2fd42cc0-2b1e-4016-8ef8-377c44a2d6e0)
    ![Screenshot from 2024-10-08 13-51-30](https://github.com/user-attachments/assets/78c2064d-80de-48c8-9191-46d562d5f2b0)
    


     </details>
3. :green_circle: `upgrade-config` w/ Skip
   
   <details>
   <summary> <i>Given</i> upgrade-config setting is set to skip there will be no prepare pods running in the cluster and the upgrade should also complete successfully
   </summary>

    
    ![Screenshot from 2024-10-08 14-38-16](https://github.com/user-attachments/assets/7a91995e-4b51-445f-b591-066641a9cfc9)
    ![Screenshot from 2024-10-08 14-31-36](https://github.com/user-attachments/assets/431fb464-e561-43d2-9d35-ad2a5bf518d4)

    


     </details>

**Comment by starbops** [comment]
Hi @irishgordo, thank you for verifying the feature. By design, the `upgrade-config` setting will not be reverted to the default value during or after the upgrade ends. Do you have the environment up and running so I can look at it?

BTW, I tried a round of same-version upgrade based on v1.4-ee6361fc-head with a 3-node cluster and the image preload option set to `skip`. The setting was not reverted to the default value after the upgrade. It's still `skip`:

```
apiVersion: harvesterhci.io/v1beta1
default: '{"imagePreloadOption":{"strategy":{"type":"sequential"}}, "restoreVM": false}'
kind: Setting
metadata:
  creationTimestamp: "2024-10-09T08:29:48Z"
  generation: 2
  name: upgrade-config
  resourceVersion: "23293"
  uid: a27b637f-22ed-4fc9-971b-2f28daec6f4c
status: {}
value: '{"imagePreloadOption":{"strategy":{"type":"skip"}}}'
```

From the `.metadata.generation`, we know that the setting object was only changed once, and it's by me.

---

> When trying to upgrade to the same version of Harvester, it seems that the harvester-network-controller-manager is stuck at deployment ready 0/1 on a 2 node cluster

Is it still an issue? I also encounter the CrashLoopBackOff of harvester-network-controller-manager :(

**Comment by irishgordo** [comment]
@starbops my appologies :sweat_smile:  - I meant to say that you can't change the "default" value.

A user could `kubectl edit setting/upgrade-config` -> changing `default:` to something like: `'{"imagePreloadOption":{"strategy":{"type":"skip"}}}'` -> but that would not persist, across reboots.

However yeah, the user can add:
```
value: '{"imagePreloadOption":{"strategy":{"type":"skip"}}}'
```
when editing `kubectl edit setting/upgrade-config` and that does persist.

So it is working as intended :smile: :+1:  
I've opened 6747 to track the other issue :smile: 

I'll go ahead and close this out -> this looks good to me on both v1.4.0-rc1 & v1.4.0-rc2 :+1:
